Apic

eim64 opened the pull request:
Adds a MADT parser and the ability to write to IO-APICs and Local APICs.

To hook the ps/2 kb interrupt you have to rewrite the correct ioapic and IOREDTBL entry.

Heres some example pseudocode that should enable keyboard interrput at 0x21
```c
    // 1 is the PIC irq for ps/2 kb
    const uint32_t gsi = g_gsi_override_table[1].gsi;
    IOAPICInfo* ioapic = get_responsible_ioapic(gsi);

    KERNEL_ASSERT(ioapic, "NO IOAPIC THAT USES KB WAS FOUND");

    const uint32_t irq = gsi - ioapic->gsi_offset;

    IOREDTBLEntry entry;
    read_redtable_entry(ioapic, irq, &entry);
    // Enable interrupt
    entry.mask = 0;

    // Fixed delivery mode
    entry.delivery_mode = 0;

    // Interrupt Vector
    entry.vector = 0x21;

    // Destination lapic
    entry.destination = g_lapic->id;

    write_redtable_entry(ioapic, irq, &entry);
```

antlilja commented on kernel/src/madt.c:
```diff
@@ -0,0 +1,103 @@
+uint64_t g_lapic_count = 0;
+
+// Local APIC's physical address
+PhysicalAddress g_lapic_physical_address = 0xFEE00000UL;
```
You shouldn't hardcode the local APIC address, it's provided in MADT header: https://wiki.osdev.org/MADT#Table_Structure

antlilja commented on kernel/src/madt.c:
```diff
@@ -0,0 +1,103 @@
+
+    while (madt < madt_end) {
+        const uint8_t type = *(uint8_t*)(madt + 0x0);
+        const uint8_t length = *(uint8_t*)(madt + 0x1);
```
Use packed structs instead of manual offsets, it's much more clear what's happening that way.

antlilja commented on kernel/src/io_apic.c:
```diff
@@ -0,0 +1,83 @@
+    *(volatile uint32_t*)(ioapic_address + 0x10) = value;
+}
+
+uint32_t read_ioapic_id(void* ioapic_address) {
```
Why not use a packed struct to read values of the IO APIC instead of a bunch of functions?

antlilja commented on kernel/include/lapic.h:
```diff
@@ -0,0 +1,151 @@
+
+// Stolen from https://docs.rs/x86/0.21.0/src/x86/xapic.rs.html#76
+#pragma region XAPIC REGISTERS
+#define XAPIC_ID 0x020
```
Use a packed struct instead of these defines, you can even mark members as const to make the compiler enforce their read only status.

antlilja reviewed (changes requested):
Some changes are required before mergin.

Don't do manual offsetting into the memory mapped IO and local APICs, use a pointer to a struct instead and let the compiler do the work for you.

antlilja commented on kernel/include/madt.h:
```diff
@@ -0,0 +1,54 @@
+extern IOAPICInfo g_found_ioapics[MAX_IOAPIC_COUNT];
+extern uint64_t g_ioapic_count;
+
+extern LAPICInfo* g_found_lapics[MAX_LAPIC_COUNT];
```
Do all these global variables need to be exposed in the header?

antlilja commented on kernel/src/io_apic.c:
```diff
@@ -0,0 +1,83 @@
+}
+
+void disable_8259() {
+    // start initialization with ICW4
```
Start sentences with capital letter

antlilja commented on kernel/include/lapic.h:
```diff
@@ -0,0 +1,83 @@
+} __attribute__((packed)) IRR;
+
+typedef struct 
+{
```
This file doesn't seem to be formatted.

antlilja commented on kernel/include/lapic.h:
```diff
@@ -0,0 +1,83 @@
+#pragma once
+#include <stdint.h>
+
+#define XVIAligned __attribute__((aligned(0x10)))
```
Macros should be all capital letters. XVIAligned doesn't really describe what it does either. But this macro doesn't really save that much space and it pollutes the global namespace with a very specific macro. You should probably just use the attribute by itself.

antlilja commented on kernel/src/io_apic.c:
```diff
@@ -0,0 +1,83 @@
+    *(volatile uint32_t*)(ioapic_address) = offset;
+
+    // return IOREGWIN
+    return *(volatile uint32_t*)(ioapic_address + 0x10);
```
You should still use a struct here instead of manual offsets.

antlilja commented on kernel/src/lapic.c:
```diff
@@ -0,0 +1,21 @@
+// This approach should work as long as all processors with kernel access use the same page tables.
+
+// Paged pointer to the physical xAPIC MMIO interface.
+volatile XAPIC* g_lapic;
```
The struct itself should be marked as volatile instead of the individual variables

antlilja reviewed (changes requested):
Maybe the different files should be merged into one? The source and header wouldn't really be that long and they interact with each other a lot.

antlilja commented on kernel/src/apic.c:
```diff
@@ -0,0 +1,259 @@
+}
+
+// IOAPIC IO
+typedef struct {
```
This struct should be defined with the others.

antlilja commented on kernel/src/apic.c:
```diff
@@ -0,0 +1,259 @@
+        }
+    }
+
+    { // Enable Local APIC
```
Don't you need to set the the enable bit in the MSR to enable the local APIC?

eim64 replied:
MSR bit is be set at boot time and cannot be disabled afterwards

antlilja replied:
Could you link the place where you read that local APIC is always enabled at boot time? Because this osdev article seems to mention hardware enabling the local APIC: https://wiki.osdev.org/APIC#Local_APIC_configuration

antlilja commented on kernel/include/apic.h:
```diff
@@ -0,0 +1,146 @@
+    const uint32_t current_count XVIALIGNED;
+    uint8_t reserved4[0x3E0 - 0x3A0] XVIALIGNED;
+    uint32_t divide_configuration XVIALIGNED;
+} __attribute__((packed)) XAPIC;
```
Why is this struct called XAPIC instead of LocalAPIC?

antlilja commented on kernel/include/apic.h:
```diff
@@ -0,0 +1,146 @@
+#include <stdint.h>
+#include <stdbool.h>
+
+#define XVIALIGNED __attribute__((aligned(0x10)))
```
This macro should be eliminated and the attribute should be used instead. 
Or you could add an ALIGN macro to utils.h which is more general purpose.

eim64 committed:
Added ioapic code

This commit will provide a struct for IOREDTBL
entries and a convenient way of writing them to an ioapic. It also comes
with a functions to find common info from the ioapic like the id, version
and irqs they use.

eim64 committed:
Added MADT Parsing that also pages IOAPICs

This makes it possible to find and use IO-APICs.
Also disables the 8259 PIC.

The g_gsi_override_table uses indexes as an 8259 PIC irq line, such that
the Global System Interrupt is easy to find when for instance hooking the
PS/2 keyboard interrupt.

To have a single struct represent an ioapic it's convenient to also store
its virtual address and irq count. to get a virtual addres you need to
page the IOAPIC, and to get the irq cont you need to then read from the
paged location. So it's very convenient to do it all when reading the MADT
rather than having additional functions that fill in those values
afterwards.

eim64 committed:
Added Local APIC support

Uses xAPIC rather than x2APIC MSR registers, because it's backward
compatible and easy to access using structs instead of having to write
each register using a wrsmr wrapper.

Note that this is quite bare bones just to get PIC interrupts working,
there are no structs etc for ICR or any of the IVTs or timers. IMO that
should be written whenever it has to be used.

antlilja commented on kernel/src/apic.c:
```diff
@@ -0,0 +1,296 @@
+        }
+
+        // Page Local APIC
+        g_lapic = (volatile LocalAPIC*)map_range(
```
You don't need volatile here because the struct is already marked as volatile

antlilja commented on kernel/src/apic.c:
```diff
@@ -0,0 +1,296 @@
+
+// IOAPIC IO
+typedef struct {
+    volatile uint32_t current_register __attribute__((aligned(0x10)));
```
You should just mark the whole struct as volatile if every member is volatile

antlilja commented on kernel/src/apic.c:
```diff
@@ -0,0 +1,296 @@
+#define APIC_BASE_MSR_ENABLE 0x800
+
+// Gets lower register offset for irq redtable entry
+#define REDTBL_OFFSET(irq) (2 * irq + 0x10)
```
You should enclose the arguments in both of these macros with parentheses.
Example:
```c
#define REDTBL_OFFSET(irq) (2 * (irq) + 0x10)
```

antlilja reviewed (approved):
Nice work!
